s3transfer/download.py

```diff
@@ -614,6 +614,38 @@ def _main(self):
         pass
 
 
+class DownloadChunkIterator(object):
+    def __init__(self, body, chunksize, response):
+        """Iterator to chunk out a downloaded S3 stream
+
+        :param body: A readable file-like object
+        :param chunksize: The amount to read each time
+        :param response: A S3 GetObject response
+        """
+        self._body = body
+        self._chunksize = chunksize
+        self._response = response
+        self._num_reads = 0
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        chunk = self._body.read(self._chunksize)
+        self._num_reads += 1
+        if chunk:
+            return chunk
+        elif self._num_reads == 1 and not self._response.get('ContentLength'):
```

Comment:
That seems fine.   The other alternative if you don't want to have to check every single time if `_num_reads` is not `1` is to just make `__iter__` a generator, check the content length header once, and if it's greater than 0 then iterate over the body.

```
def __iter__(self):
    if self.headers['Content-Length'] == 0:
        yield b''
    else:
        for chunk in iter(self._body, lambda: ....):
            yield chunk
```